### PR TITLE
Add support for Spark cluster.

### PR DESCRIPTION
**Pull request by vitillo**

### CODE REVIEW

File: http/analysis-service/server.py [hunk 1]
```diff
@@ -846,6 +849,135 @@ def kill(instance_id):
...
+
+    # Check required file
+    if not request.files['public-ssh-key']:
+        errors['code-tarball'] = "Public key file is required"
+
+    # Bug 961200: Check that a proper OpenSSH public key was uploaded.
+    # It should start with "ssh-rsa AAAAB3"
+    pubkey = request.files['public-ssh-key'].read()
+    if not pubkey.startswith("ssh-rsa AAAAB3") and not pubkey.startswith("ssh-dss AAAAB3"):
+        errors['public-ssh-key'] = "Supplied file does not appear to be a valid OpenSSH public key."
```
Comment (mreid-moz) [inline http/analysis-service/server.py]: Please extract a common function for validating the public key (here and in `spawn_worker_instance()`)

File: http/analysis-service/server.py [hunk 2]
```diff
@@ -846,6 +849,135 @@ def kill(instance_id):
...
+
+    if errors:
+        return cluster_get_params(errors, request.form)
+
+    # Create EMR cluster
+    n_workers = int(request.form["n-workers"])
```
Comment (mreid-moz) [inline http/analysis-service/server.py]: Check `n_workers > 0`

File: http/analysis-service/config.py [hunk 3]
```diff
@@ -4,6 +4,15 @@
 # AWS EC2 configuration
 AWS_REGION         = 'us-west-2'
 INSTANCE_TYPE      = 'c3.4xlarge'
+
+# EMR configuration
+# Master and slave instance types should be the same as the telemetry
+# setup bootstrap action depends on it to autotune the cluster.
+MASTER_INSTANCE_TYPE = 'c3.4xlarge'
+SLAVE_INSTANCE_TYPE = 'c3.4xlarge'
```
Comment (mreid-moz) [inline http/analysis-service/config.py]: Maybe just do `MASTER_INSTANCE_TYPE = INSTANCE_TYPE` if they must be the same. Or just use the `INSTANCE_TYPE` setting directly when launching the cluster.

File: http/analysis-service/server.py [hunk 4]
```diff
@@ -846,6 +849,135 @@ def kill(instance_id):
...
+def cluster_monitor(jobflow_id):
+    # Check that the user logged in is also authorized to do this
+    if not current_user.is_authorized():
+        return login_manager.unauthorized()
+
+    try:
+        jobflow = emr.describe_jobflow(jobflow_id)
+    except:
+        return "No such cluster: {}".format(jobflow_id), 404
+
```
Comment (mreid-moz) [inline http/analysis-service/server.py]: Users should probably only be able to monitor their own clusters. Check the `Owner` tag to make sure it matches the logged in user's email.

File: http/analysis-service/server.py [hunk 5]
```diff
@@ -846,6 +849,135 @@ def kill(instance_id):
...
+    # Terminate cluster
+    emr.terminate_jobflow(jobflow_id)
```
Comment (mreid-moz) [inline http/analysis-service/server.py]: Users should **definitely** only be able to kill their own clusters :) Check the `Owner` tag to make sure it matches the logged in user's email.

File: http/analysis-service/templates/index.html [hunk 6]
```diff
@@ -2,9 +2,21 @@
...
 <div class="row">
- <div class="col-md-6">
+ <div class="col-md-4">
+  <h2>Launch a Spark Cluster</h2>
+  <p>Launch a Spark cluster in the cloud and use it for custom data analysis.
+  The cluster will be available for 24 hours, then it will be automatically
+  terminated.</p>
+  <p class="lead">
+   <a href="{{ url_for('cluster_get_params') }}">
+    <img width="48" src="{{ url_for('static', filename='worker.png') }}" />
```
Comment (mreid-moz) [inline http/analysis-service/templates/index.html]: s/worker.png/cluster.png/

**Comment by vitillo** [comment]
I have updated the PR with your comments.

File: http/analysis-service/server.py [hunk 7]
```diff
@@ -846,6 +859,146 @@ def kill(instance_id):
...
+    install_spark_bootstrap = BootstrapAction('Install Spark',
+                                              's3://support.elasticmapreduce/spark/install-spark',
+                                              ['-v', app.config['SPARK_VERSION']])
+
+    setup_telemetry_bootstrap = BootstrapAction('Setup Telemetry',
+                                                's3://telemetry-spark-emr/telemetry.sh',
+                                                ['--public-key', pubkey])
+
+    jobflow_id = emr.run_jobflow(name = request.form['token'],
+                              ec2_keyname = 'mozilla_vitillo',
```
Comment (mreid-moz) [inline http/analysis-service/server.py]: We should eventually not use our own keys, but I see I already did the same thing for the worker instances, so let's not worry about it for now.

**Comment by mreid-moz** [comment]
Looks good now, thanks!

**Pull request merged**